python/demo/gsoc2022/demo_scattering_boundary_conditions.py

```diff
@@ -0,0 +1,448 @@
+# and the horizontal axis $\hat{\mathbf{u}}_x$,
+# we can write more explicitly:
+#
+# $$
+# \mathbf{E}_b = -\sin\theta e^{j (k_xx+k_yy)}\hat{\mathbf{u}}_x
+# + \cos\theta e^{j (k_xx+k_yy)}\hat{\mathbf{u}}_y
+# $$
+#
+# The Maxwell's equation for scattering problems takes the following form:
+#
+# $$
+# \textrm{on } \Omega: \quad-\nabla \times
+# \nabla \times \mathbf{E}_s+\varepsilon_{r} k_{0}^{2} \mathbf{E}_s
+# +k_{0}^{2}\left(\varepsilon_{r}-\varepsilon_{b}\right)
+# \mathbf{E}_{\mathrm{b}}=0,
+# $$
+#
+# where $k_0 = 2\pi/\lambda_0$ is the vacuum wavevector of the background
+# field, $\varepsilon_b$ is the background relative permittivity and
+# $\varepsilon_r$ is the relative permittivity as a function of space,
+# i.e.:
+#
+# $$
+# \varepsilon_r = \begin{cases}
+# \varepsilon_m & \textrm{on }\Omega_m \\
+# \varepsilon_b & \textrm{on }\Omega_b
+# \end{cases}
+# $$
+#
+# with $\varepsilon_m$ being the relative permittivity of the matallic
+# wire. As reference values, we will consider $\lambda_0 = 400\textrm{nm}$
+# (violet light), $\varepsilon_b = 1.33^2$ (relative permittivity of water),
+# and $\varepsilon_m = -1.0782 + 5.8089\textrm{j}$ (relative permittivity of
+# gold at $400\textrm{nm}$ (ref)).
+#
+# To make the system determined, we need to add some boundary conditions
+# on $\partial \Omega$. A common approach is the use of scattering
+# boundary conditions (ref), which make the boundary transparent for
+# $\mathbf{E}_s$, allowing us to restric the computational boundary
+# to a finite $\Omega$ domain. The first-order boundary conditions
+# in the 2D case take the following form:
+#
+# $$
+# \textrm{on } \partial \Omega: \quad \mathbf{n} \times
+# \nabla \times \mathbf{E}_s+\left(j k_{0}n_b + \frac{1}{2r}
+# \right) \mathbf{n} \times \mathbf{E}_s
+# \times \mathbf{n}=0
+# $$
+#
+# with $n_b = \sqrt{\varepsilon_b}$ being the background refractive
+# index, $\mathbf{n}$ being the normal versor to $\partial \Omega$,
+# and $r = \sqrt{(x-x_s)^2 + (y-y_s)^2}$ being the distance of the
+# $(x, y)$ point on $\partial\Omega$ from the wire centered in
+# $(x_s, y_s)$. In our case we will consider
+# the wire centered in the origin of our mesh, and therefore $r =
+# \sqrt{x^2 + y^2}$.
+#
+# Now we need to find the weak form of the problem. First of all,
+# we need to take the inner products of the equations with a
+# complex test function $\mathbf{V}$, and then we need to integrate
+# the terms over the corresponding domains:
+#
+# $$
+# \begin{align}
+# & \iint_{\Omega}-\nabla \times \nabla \times \mathbf{E}_s \cdot
+# \bar{\mathbf{V}}+\varepsilon_{r} k_{0}^{2} \mathbf{E}_s \cdot
+# \bar{\mathbf{V}}+k_{0}^{2}\left(\varepsilon_{r}-\varepsilon_b\right)
+# \mathbf{E}_b \cdot \bar{\mathbf{V}} d S \\ +& \int_{\partial \Omega}
+# \mathbf{n} \times \nabla \times \mathbf{E}_s \cdot \bar{\mathbf{V}}
+# +\left(j n_bk_{0}+\frac{1}{2r}\right) \mathbf{n} \times \mathbf{E}_s
+# \times \mathbf{n} \cdot \bar{\mathbf{V}} d l=0
+# \end{align}
+# $$
```

Comment:
```suggestion
# $$
# \begin{align}
# & \int_{\Omega}-\nabla \times( \nabla \times \mathbf{E}_s) \cdot
# \mathbf{v}+\varepsilon_{r} k_{0}^{2} \mathbf{E}_s \cdot
# \mathbf{v}+k_{0}^{2}\left(\varepsilon_{r}-\varepsilon_b\right)
# \mathbf{E}_b \cdot \mathbf{v}~\mathrm{d}x \\ +& \int_{\partial \Omega}
# (\mathbf{n} \times \nabla \times \mathbf{E}_s) \cdot \mathbf{v}
# +\left(j n_bk_{0}+\frac{1}{2r}\right) (\mathbf{n} \times \mathbf{E}_s
# \times \mathbf{n}) \cdot \mathbf{v}~\mathrm{d}s=0
# \end{align}
# $$
```
To align with DOLFINx notation, I suggest we use `v` as a `TestFunction`, `dx` for volume integrals and `ds` for surface integrals